File: backend/services/commentService.js
const { query } = require('../config/db');

class CommentService {
  async getCommentsByPostId(postId, currentUserId = null) {
    const [rows] = await query(
      `SELECT c.id, c.post_id, c.user_id, c.parent_id, c.content, c.created_at, c.updated_at,
              u.username, u.display_name, u.avatar_url,
              (SELECT COUNT(*) FROM comment_likes WHERE comment_id = c.id) AS likes_count,
              (SELECT COUNT(*) FROM comment_likes WHERE comment_id = c.id AND user_id = ?) AS is_liked
       FROM comments c
       JOIN users u ON u.id = c.user_id
       WHERE c.post_id = ? AND u.is_suspended = 0
       ORDER BY c.created_at ASC`,
      [currentUserId || 0, postId]
    );

    const commentMap = {};
    const rootComments = [];

    // First pass: instantiate comment items with empty replies
    for (const r of rows) {
      r.is_liked = Boolean(r.is_liked);
      r.replies = [];
      commentMap[r.id] = r;
    }

    // Second pass: nest replies into parents
    for (const r of rows) {
      if (r.parent_id && commentMap[r.parent_id]) {
        commentMap[r.parent_id].replies.push(r);
      } else if (!r.parent_id) {
        rootComments.push(r);
      }
    }

    return rootComments;
  }

  async createComment(userId, postId, content, parentId = null) {
    if (!content || !content.trim()) {
      throw new Error('Comment content cannot be empty.');
    }

    // Validate post
    const [postRows] = await query('SELECT id, user_id FROM posts WHERE id = ?', [postId]);
    if (postRows.length === 0) throw new Error('Post not found.');
    const postOwnerId = postRows[0].user_id;

    // Validate parent comment if provided
    let parentCommentOwnerId = null;
    if (parentId) {
      const [parentRows] = await query('SELECT id, user_id FROM comments WHERE id = ? AND post_id = ?', [parentId, postId]);
      if (parentRows.length === 0) throw new Error('Parent comment not found.');
      parentCommentOwnerId = parentRows[0].user_id;
    }

    const [res] = await query(
      `INSERT INTO comments (post_id, user_id, parent_id, content) VALUES (?, ?, ?, ?)`,
      [postId, userId, parentId || null, content.trim()]
    );
    const commentId = res.insertId;

    // Send Notification
    if (parentId && parentCommentOwnerId && Number(parentCommentOwnerId) !== Number(userId)) {
      await query(
        'INSERT INTO notifications (recipient_id, actor_id, type, entity_id, entity_type, message) VALUES (?, ?, "REPLY", ?, "POST", "replied to your comment.")',
        [parentCommentOwnerId, userId, postId]
      );
    } else if (Number(postOwnerId) !== Number(userId)) {
      await query(
        'INSERT INTO notifications (recipient_id, actor_id, type, entity_id, entity_type, message) VALUES (?, ?, "COMMENT", ?, "POST", "commented on your post.")',
        [postOwnerId, userId, postId]
      );
    }

    // Fetch created comment
    const [createdRows] = await query(
      `SELECT c.id, c.post_id, c.user_id, c.parent_id, c.content, c.created_at,
              u.username, u.display_name, u.avatar_url,
              0 AS likes_count, 0 AS is_liked
       FROM comments c
       JOIN users u ON u.id = c.user_id
       WHERE c.id = ?`,
      [commentId]
    );

    const newComment = createdRows[0];
    newComment.replies = [];
    newComment.is_liked = false;
    return newComment;
  }

  async deleteComment(commentId, userId, userRole = 'USER') {
    const [rows] = await query(
      `SELECT c.id, c.user_id, p.user_id AS post_author_id
       FROM comments c
       JOIN posts p ON p.id = c.post_id
       WHERE c.id = ?`,
      [commentId]
    );

    if (rows.length === 0) throw new Error('Comment not found.');

    const isCommentAuthor = Number(rows[0].user_id) === Number(userId);
    const isPostAuthor = Number(rows[0].post_author_id) === Number(userId);
    const isAdmin = userRole === 'ADMIN';

    if (!isCommentAuthor && !isPostAuthor && !isAdmin) {
      throw new Error('Unauthorized to delete this comment.');
    }

    await query('DELETE FROM comments WHERE id = ?', [commentId]);
    return { success: true, message: 'Comment deleted.' };
  }

  async toggleCommentLike(userId, commentId) {
    const [existing] = await query('SELECT id FROM comment_likes WHERE user_id = ? AND comment_id = ?', [userId, commentId]);
    let isLiked = false;

    if (existing.length > 0) {
      await query('DELETE FROM comment_likes WHERE user_id = ? AND comment_id = ?', [userId, commentId]);
      isLiked = false;
    } else {
      await query('INSERT INTO comment_likes (user_id, comment_id) VALUES (?, ?)', [userId, commentId]);
      isLiked = true;
    }

    const [countRows] = await query('SELECT COUNT(*) AS count FROM comment_likes WHERE comment_id = ?', [commentId]);
    return { is_liked: isLiked, likes_count: countRows[0].count, comment_id: Number(commentId) };
  }
}

module.exports = new CommentService();

File: backend-java/src/main/java/com/socialmedia/service/CommentService.java
package com.socialmedia.service;

import com.socialmedia.dto.Dtos.*;
import com.socialmedia.entity.Comment;
import com.socialmedia.entity.CommentLike;
import com.socialmedia.entity.Notification;
import com.socialmedia.entity.Post;
import com.socialmedia.entity.User;
import com.socialmedia.repository.PostRepository;
import com.socialmedia.repository.CommentLikeRepository;
import com.socialmedia.repository.CommentRepository;
import com.socialmedia.repository.NotificationRepository;
import com.socialmedia.repository.UserRepository;
import com.socialmedia.websocket.WebSocketConfig.ChatWebSocketHandler;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
public class CommentService {

    @Autowired
    private CommentRepository commentRepository;

    @Autowired
    private CommentLikeRepository commentLikeRepository;

    @Autowired
    private PostRepository postRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private NotificationRepository notificationRepository;

    @Autowired
    private ChatWebSocketHandler wsHandler;

    @Transactional(readOnly = true)
    public List<CommentResponse> getComments(Long postId, Long currentUserId) {
        Post post = postRepository.findById(postId)
                .orElseThrow(() -> new IllegalArgumentException("Post not found"));

        return commentRepository.findByPostAndParentCommentIsNullOrderByCreatedAtAsc(post).stream()
                .map(c -> mapToCommentResponse(c, currentUserId))
                .collect(Collectors.toList());
    }

    @Transactional
    public CommentResponse createComment(Long userId, Long postId, CreateCommentRequest req) {
        if (req.getContent() == null || req.getContent().isBlank()) {
            throw new IllegalArgumentException("Comment content cannot be empty.");
        }

        User user = userRepository.findById(userId).orElseThrow();
        Post post = postRepository.findById(postId).orElseThrow(() -> new IllegalArgumentException("Post not found"));

        Comment parent = null;
        if (req.getParentId() != null) {
            parent = commentRepository.findById(req.getParentId()).orElse(null);
        }

        Comment comment = Comment.builder()
                .user(user)
                .post(post)
                .parentComment(parent)
                .content(req.getContent().trim())
                .build();

        comment = commentRepository.save(comment);

        // Notifications
        if (parent != null && !parent.getUser().getId().equals(userId)) {
            Notification notif = Notification.builder()
                    .recipient(parent.getUser())
                    .actor(user)
                    .type(Notification.Type.REPLY)
                    .entityId(postId)
                    .entityType("POST")
                    .message("replied to your comment.")
                    .build();
            notificationRepository.save(notif);
            wsHandler.sendToUser(parent.getUser().getId(), Map.of("action", "NEW_NOTIFICATION", "message", user.getDisplayName() + " replied to your comment."));
        } else if (!post.getUser().getId().equals(userId)) {
            Notification notif = Notification.builder()
                    .recipient(post.getUser())
                    .actor(user)
                    .type(Notification.Type.COMMENT)
                    .entityId(postId)
                    .entityType("POST")
                    .message("commented on your post.")
                    .build();
            notificationRepository.save(notif);
            wsHandler.sendToUser(post.getUser().getId(), Map.of("action", "NEW_NOTIFICATION", "message", user.getDisplayName() + " commented on your post."));
        }

        return mapToCommentResponse(comment, userId);
    }

    @Transactional
    public void deleteComment(Long commentId, Long userId, boolean isAdmin) {
        Comment comment = commentRepository.findById(commentId)
                .orElseThrow(() -> new IllegalArgumentException("Comment not found"));

        boolean isAuthor = comment.getUser().getId().equals(userId);
        boolean isPostAuthor = comment.getPost().getUser().getId().equals(userId);

        if (!isAuthor && !isPostAuthor && !isAdmin) {
            throw new IllegalArgumentException("Unauthorized to delete this comment.");
        }

        commentRepository.delete(comment);
    }

    @Transactional
    public Map<String, Object> toggleCommentLike(Long userId, Long commentId) {
        User user = userRepository.findById(userId).orElseThrow();
        Comment comment = commentRepository.findById(commentId).orElseThrow();

        var existing = commentLikeRepository.findByUserAndComment(user, comment);
        boolean isLiked;
        if (existing.isPresent()) {
            commentLikeRepository.delete(existing.get());
            isLiked = false;
        } else {
            commentLikeRepository.save(CommentLike.builder().user(user).comment(comment).build());
            isLiked = true;
        }

        long likesCount = commentLikeRepository.countByComment(comment);
        return Map.of("is_liked", isLiked, "likes_count", likesCount, "comment_id", commentId);
    }

    private CommentResponse mapToCommentResponse(Comment comment, Long currentUserId) {
        long likesCount = commentLikeRepository.countByComment(comment);
        boolean isLiked = false;
        if (currentUserId != null) {
            User currentUser = userRepository.findById(currentUserId).orElse(null);
            if (currentUser != null) {
                isLiked = commentLikeRepository.existsByUserAndComment(currentUser, comment);
            }
        }

        List<CommentResponse> replies = commentRepository.findByParentCommentOrderByCreatedAtAsc(comment).stream()
                .map(r -> mapToCommentResponse(r, currentUserId))
                .collect(Collectors.toList());

        return CommentResponse.builder()
                .id(comment.getId())
                .postId(comment.getPost().getId())
                .userId(comment.getUser().getId())
                .username(comment.getUser().getUsername())
                .displayName(comment.getUser().getDisplayName())
                .avatarUrl(comment.getUser().getAvatarUrl())
                .parentId(comment.getParentComment() != null ? comment.getParentComment().getId() : null)
                .content(comment.getContent())
                .likesCount(likesCount)
                .isLiked(isLiked)
                .replies(replies)
                .createdAt(comment.getCreatedAt())
                .build();
    }
}
